packages/calypso-e2e/src/lib/flows/new-post-flow.ts

```diff
@@ -15,6 +18,28 @@ export class NewPostFlow {
 	 */
 	constructor( page: Page ) {
 		this.page = page;
+		this.isLoggedIn = false;
+	}
+
+	/**
+	 * Starts a new post by directly going to the default site of the given
+	 * account. Logs in if necessary.
+	 *
+	 * @param user The username of the account to start the post with
+	 * @returns An instance of the GutenbergEditorPage class
+	 */
+	async startImmediately( user: string ): Promise< GutenbergEditorPage > {
```

Comment:
I'm assuming this is meant to be more of a 'shortcut' aimed at reducing the boilerplate that exists for each of our e2e specs.

Boilerplate:

```
		it( 'Log in', async function () {
			const loginPage = new LoginPage( page );
			await loginPage.login( { account: postingUser } );
		} );

		it( 'Start new post', async function () {
			const newPostFlow = new NewPostFlow( page );
			await newPostFlow.newPostFromNavbar();
		} );
```

Assuming my understanding above is correct -while I like the idea, I am not sure if the current implementation of the `startImmediately` method is the answer here.

My opinion is that by allowing other flows to also handle login, it gradually leads to scope creep of classes, making it difficult to refactor or change behaviors. In this case, `NewPostFlow` would end up being coupled with how `LoginPage` is implemented, to a certain degree. 

Would a pared-back method that expects that user has completed Login be a sufficient alternative?

Reply:
> I'm assuming this is meant to be more of a 'shortcut' aimed at reducing the boilerplate that exists for each of our e2e specs

Yes, but mainly it provides a separation of concerns where our editor tests are no longer dependant on the WP.com homepage. It should also get us slightly better numbers in terms of tests duration :)

> My opinion is that by allowing other flows to also handle login, it gradually leads to scope creep of classes, making it difficult to refactor or change behaviors. In this case, NewPostFlow would end up being coupled with how LoginPage is implemented, to a certain degree.

That's a valid concern and something that I actually wanted to discuss! 😄 Currently, the `LoginPage` model looks more like a `Flow` model to me, as it contains complete scenario methods like `login`, `signup`, `requestMagicLink`, etc. In general, I believe that the `Page` models should contain per-element user actions, that can then be utilized by various `Flow` models. For example, IMO the `LoginPage` should consist of the following actions: 
- `fillUsername`,
- `fillPassword`,
- `clickContinueButton`,
- `clickLogInButton`,
- `clickContinueWithAppleButton`,
- `clickEmailMeALoginLinkButton`, etc.

...which would then be used e.g. within the `LoginFlow` model to build scenarios like `logIn`, `signUp` (both of which could also be using a `LandingPage` model!), or `requestAMagicLink`.

Going back to the original concern of coupling a `Page` model implementation with a `Flow` model: you can see how that could be avoided if the `Page` (`LoginPage` in this case) consisted only of singular user actions (click*, fill*, etc.). Having a `Page` model built in such a fashion, there would be no need to actually add anything there as it would already contain all the action methods we need.

Having said that, I agree that this direction of the coupling is not a good practice, but at the same time, I think the reason I had to create the extra `fillAndSubmit` scenario was that the expected Page methods, like `fillUsername`, `fillPassword`, etc. weren't available. 

What do you think? Do you think that it would be a good idea to separate `LoginPage` and `LoginFlow` model concerns in the way I described above?

Reply:
> Do you think that it would be a good idea to separate LoginPage and LoginFlow model concerns in the way I described above?

Interesting you say this - in https://github.com/Automattic/wp-calypso/pull/56482 I folded the functionality of `LoginFlow` into `LoginPage` because I felt there was not much to be gained by distinguishing between flows and pages for `calypso-e2e`.

Separating Flows and Pages would inevitably lead to duplication of code to some degree (like was seen with LoginPage/LoginFlow). Eliminating this duplication of code was my primary intent in merging the two together.

With that said, I can still understand and justify both perspectives. 

Here's a thought; what if the contents of `LoginPage.baseflow` was refactored out to individual methods (like `fillUsername`, `clickLoginButton`, etc.)? Doing so, it would allow you to string together only the necessary calls, minimizing the overhead.

Also, keep in mind that with `SAVE_AUTH_COOKIES` env var enabled in CI, the login step is not actually executed for users that fall under [this list](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/lib/jest/globalSetup.ts#L37). For these users, the test begins by first landing in the dashboard. I think it is also possible to do a `page.goto(<new_post_url>)` and it would load.

Reply:
All this is to say, if it helps you I would be totally up for extracting each foundational step of LoginPage out into its own standalone method.

I am less interested in re-splitting LoginPage out to LoginPage and LoginFlow for the reasons cited above, but am not 100% dead set against it.

Reply:
> Here's a thought; what if the contents of LoginPage.baseflow was refactored out to individual methods (like fillUsername, clickLoginButton, etc.)? Doing so, it would allow you to string together only the necessary calls, minimizing the overhead.

Sounds good! I'll work on a PR for that.

> Also, keep in mind that with SAVE_AUTH_COOKIES env var enabled in CI, the login step is not actually executed for users that fall under this list. For these users, the test begins by first landing in the dashboard. I think it is also possible to do a page.goto(<new_post_url>) and it would load.

👍 

---

I suggest we ship this one with the extra `fillAndSubmit` method and I'll work on improving it in a follow-up PR. How does that sound?

Reply:
Agreed, let's iterate later as required.